File: wingetui/PackageManagers/PackageClasses.py
if __name__ == "__main__":
    # WingetUI cannot be run directly from this file, it must be run by importing the wingetui module 
    print("redirecting...")
    import subprocess, os, sys
    sys.exit(subprocess.run(["cmd", "/C", "python", "-m", "wingetui"], shell=True, cwd=os.path.dirname(__file__).split("wingetui")[0]).returncode)
    
    from Interface.CustomWidgets.SpecificWidgets import PackageItem, InstalledPackageItem, UpgradablePackage
    # TODO: Check this import hierarchy


import os
import subprocess
from PySide6.QtCore import *
from PySide6.QtGui import *
from PySide6.QtWidgets import *
from urllib.request import urlopen

import wingetui.Core.Globals as Globals
from wingetui.Core.Tools import *
from wingetui.Core.Tools import _, blueColor


class Package():
    Name: str = ""
    Id: str = ""
    UniqueId: str = ""
    Version: str = ""
    Source: str = ""
    PackageItem: 'PackageItem' = None
    PackageManager: 'PackageManagerModule' = None

    def __init__(self, Name: str, Id: str, Version: str, Source: str, PackageManager: 'PackageManagerModule'):
        self.Name = Name
        self.Id = Id
        self.Version = Version
        self.Source = Source
        self.PackageManager = PackageManager
        self.UniqueId = self.PackageManager.NAME + "\\" + self.Id + "\\" + self.Version

    def isWinget(self) -> bool:
        return "winget" in self.Source.lower()

    def isScoop(self) -> bool:
        return "scoop" in self.Source.lower()

    def isChocolatey(self) -> bool:
        return self.Source.lower() == "chocolatey"

    def getIconId(self) -> str:
        """
        Returns the normalized id.
        """
        iconId = self.Id.lower()
        if self.isWinget():
            iconId = ".".join(iconId.split(".")[1:])
        elif self.isChocolatey():
            iconId = iconId.replace(".install", "").replace(".portable", "")
        elif self.isScoop():
            iconId = iconId.split("/")[-1]
        return iconId.replace(" ", "-").replace("_", "-").replace(".", "-")

    def getPackageIconUrl(self) -> str:
        try:
            iconId = self.getIconId()
            try:
                iconUrl = Globals.packageMeta["icons_and_screenshots"][iconId]["icon"]
                if iconUrl.strip() == "":
                    raise KeyError("Key found but content was empty")
            except KeyError:
                if "Net" in self.Source:
                    iconUrl = f"https://api.nuget.org/v3-flatcontainer/{self.Id}/{self.Version}/icon"
                elif "Chocolatey" in self.Source:
                    iconUrl = f"https://community.chocolatey.org/content/packageimages/{self.Id}.{self.Version}.png"
                else:
                    iconUrl = ""
        except Exception as e:
            report(e)
            iconUrl = ""
        return iconUrl

    def getPackageIcon(self) -> str:
        """
        Returns a string containing the complete path to a local PNG file containing this package's icon.
        If needed, the icon will be downloaded. 
        """
        try:
            iconId = self.getIconId()
            iconPath = os.path.join(ICON_DIR, f"{iconId}.icon.png")
            if not os.path.exists(ICON_DIR):
                os.makedirs(ICON_DIR)
            if not os.path.exists(iconPath):
                iconUrl = self.getPackageIconUrl()
                print("🔵 Found icon: ", iconUrl)
                if iconUrl:
                    iconData = urlopen(iconUrl).read()
                    with open(iconPath, "wb") as f:
                        f.write(iconData)
                else:
                    print("🟡 Icon url empty")
                    raise KeyError(f"{iconUrl} was empty")
            else:
                print(f"🔵 Found cached image in {iconPath}")
            return iconPath
        except KeyError:
            print(f"🟡 Icon {iconId} not found in json (KeyError)")
            return ""
        except Exception as e:
            report(e)
            return ""

    def getSourceIcon(self) -> QIcon:
        """
        Returns the icon of the PackageManager in the QIcon format
        """
        return self.PackageManager.getIcon(self.Source)

    def isManager(self, manager: 'PackageManagerModule') -> bool:
        """
        Check if the package manager equals the given package manager
        """
        return manager == self.PackageManager

    def getFloatVersion(self) -> float:
        """
        Returns a float number representing the version of the package.
        """
        newVer = ""
        dotAdded = False
        for char in self.Version:
            if char in "0123456789":
                newVer += char
            elif char == ".":
                if not dotAdded:
                    newVer += "."
                    dotAdded = True
        if newVer and newVer != ".":
            strVer = f"{float(newVer):040.10f}"
        else:
            strVer = f"{0.0:040.10f}"
        return strVer

    def isTheSameAs(self, package: 'Package'):
        return self.Id == package.Id and self.Name == package.Name and package.Source == package.Source and package.PackageManager == package.PackageManager

    def __str__(self) -> str:
        return f"<Package: {self.Name};{self.Id};{self.Version};{self.Source};{self.PackageManager};{self.PackageItem}>"
        
    def AddToIgnoredUpdates(self, version: str = "*"):
        """
        Add a package to the ignored package updates list.
        If the parameter version is given, the given version will be ignored. Otherwise, all versions will.
        """
        ENTRY = self.Source.lower().split(":")[0] + "\\" + self.Id
        ignoredPackages = GetJsonSettings("IgnoredPackageUpdates")
        ignoredPackages[ENTRY] = version
        SetJsonSettings("IgnoredPackageUpdates", ignoredPackages)
        
        if self.PackageItem is not None:
            InstalledItem = self.PackageItem.getInstalledPackageItem()
            if InstalledItem:
                InstalledItem.setTag(InstalledItem.Tag.Pinned)
            UpgradableItem = self.PackageItem.getUpdatesPackageItem()
            if UpgradableItem:
                UpgradableItem.removeFromList()


    def RemoveFromIgnoredUpdates(self):
        """
        Remove a package (if present) from the ignored packages list.
        """
        ENTRY = self.Source.lower().split(":")[0] + "\\" + self.Id
        ignoredPackages = GetJsonSettings("IgnoredPackageUpdates")
        if ENTRY in ignoredPackages.keys():
            del ignoredPackages[ENTRY]
                 
        if self.PackageItem is not None:
            InstalledItem = self.PackageItem.getInstalledPackageItem()
            if InstalledItem:
                InstalledItem.setTag(InstalledItem.Tag.Default)
        
        SetJsonSettings("IgnoredPackageUpdates", ignoredPackages)
        
    def HasUpdatesIgnored(self, version: str = "*") -> bool:
        """
        Return if a package is being ignored for the given version.
        If version is not given (or the wildcard "*" is used), all the versions will be checked.
        """
        ENTRY = self.Source.lower().split(":")[0] + "\\" + self.Id
        ignoredPackages = GetJsonSettings("IgnoredPackageUpdates")
        if ENTRY in ignoredPackages.keys():
            if ignoredPackages[ENTRY] == "*":
                return True # Will take into account the case where the package has been ignored for all versions but a specific version is checked.
            elif ignoredPackages[ENTRY] == version:
                return True
        return False

    def GetIgnoredUpatesVersion(self) -> str:
        """
        Returns the version for which a package has been ignored. Will return the wildcard "*" if all the versions are ignored.
        If the package is not ignored returns an empty string.
        """
        ENTRY = self.Source.lower().split(":")[0] + "\\" + self.Id
        ignoredPackages = GetJsonSettings("IgnoredPackageUpdates")
        if ENTRY in ignoredPackages.keys():
            return ignoredPackages[ENTRY]
        return ""

    def getDiscoverPackage(self) -> 'Package':
        if self.PackageItem:
            # This function is more efficient if wanting to find the same item
            AvailablePackageItem = self.PackageItem.getDiscoverPackageItem()
            if AvailablePackageItem:
                return AvailablePackageItem.Package
        if self.Id in Globals.discover.IdPackageReference:
            package: Package = Globals.discover.IdPackageReference[self.Id]
            if package.Source == self.Source:
                return package
        return None

    def getUpdatesPackage(self) -> 'UpgradablePackage':
        if self.PackageItem:
            # This function is more efficient if wanting to find the same item
            UpgrdablePackageItem = self.PackageItem.getUpdatesPackageItem()
            if UpgrdablePackageItem:
                return UpgrdablePackageItem.Package
        if self.Id in Globals.updates.IdPackageReference:
            package: UpgradablePackage = Globals.updates.IdPackageReference[self.Id]
            if package.Source == self.Source:
                return package
        return None

    def getInstalledPackage(self) -> 'Package':
        if self.PackageItem:
            # This function is more efficient if wanting to find the same item
            InstalledPackageItem = self.PackageItem.getInstalledPackageItem()
            if InstalledPackageItem:
                return InstalledPackageItem.Package
        if self.Id in Globals.uninstall.IdPackageReference:
            package: Package = Globals.uninstall.IdPackageReference[self.Id]
            if self.Source in package.Source:  # Allow "Scoop" packages to be detected as "Scoop: bucket" sources
                return package
        return None
    
    def getAllCorrespondingInstalledPackages(self) -> 'list[Package]':
        matches = []
        if self.Id in Globals.uninstall.IdPackageReference:
            package: Package = Globals.uninstall.IdPackageReference[self.Id]
            if self.Source in package.Source:  # Allow "Scoop" packages to be detected as "Scoop: bucket" sources
                matches.append(package)
        return matches



class UpgradablePackage(Package):
    NewVersion = ""
    NewPackage: Package = None

    def __init__(self, Name: str, Id: str, InstalledVersion: str, NewVersion: str, Source: str, PackageManager: 'PackageManagerModule'):
        super().__init__(Name, Id, InstalledVersion, Source, PackageManager)
        self.NewVersion = NewVersion
        self.NewPackage = Package(Name, Id, NewVersion, Source, PackageManager)

class PackageDetails(Package):
    Name: str = ""
    Id: str = ""
    Version: str = ""
    NewVersion: str = ""
    Source: str = ""
    PackageObject: Package = None
    Publisher: str = _("Not available")
    Author: str = _("Not available")
    Description: str = _("Not available")
    HomepageURL: str = _("Not available")
    License: str = ""
    LicenseURL: str = ""
    InstallerURL: str = _("Not available")
    InstallerHash: str = _("Not available")
    InstallerSize: int = 0  # In Megabytes
    InstallerType: str = _("Not available")
    ManifestUrl: str = _("Not available")
    UpdateDate: str = _("Not available")
    ReleaseNotes: str = _("Not available")
    ReleaseNotesUrl: str = _("Not available")
    Versions: list[str] = []
    Architectures: list[str] = []
    Scopes: list[str] = []
    Tags: list[str] = []

    def __init__(self, package: Package):
        self.Name = package.Name
        self.Id = package.Id
        self.Version = package.Version
        self.Source = package.Source
        self.PackageObject = package
        if type(package) is UpgradablePackage:
            self.NewVersion = package.NewVersion

    def asUrl(self, url: str) -> str:
        return f"<a href='{url}' style='color:{blueColor}'>{url}</a>" if "://" in url else url


class InstallationOptions():
    SkipHashCheck: bool = False
    InteractiveInstallation: bool = False
    RunAsAdministrator: bool = False
    Version: str = ""
    Architecture: str = ""
    InstallationScope: str = ""
    CustomParameters: list[str] = []
    RemoveDataOnUninstall: bool = False
    PreRelease: bool = False
    CustomInstallLocation: str = ""

    Package: 'Package' = None
    __save_file_name: str = "Unknown.Unknown.InstallationOptions"
    __data_to_save: list[str] = [
            "SkipHashCheck",
            "InteractiveInstallation",
            "RunAsAdministrator",
            "Architecture",
            "InstallationScope",
            "CustomParameters",
            "PreRelease",
            "CustomInstallLocation"
        ]

    def __init__(self, package: 'Package', reset: bool = False):
        self.Package = package
        self.__save_file_name = self.Package.PackageManager.NAME.replace(" ", "").replace(".", "") + "." +  self.Package.Id
        if not reset:
            self.LoadOptionsFromDisk()
            
    def ToJson(self) -> dict:
        optionsToSave = {}
        for entry in self.__data_to_save:
            optionsToSave[entry] = getattr(self, entry)
        return optionsToSave
    
    def LoadFromJson(self, data: dict):
        for entry in self.__data_to_save:
            if entry in data.keys():
                setattr(self, entry, data[entry])

    def SaveOptionsToDisk(self):
        """
        Save current installation options to disk
        """
        SetJsonSettings(Name=self.__save_file_name, Data=self.ToJson(), Scope="InstallationOptions")

    def LoadOptionsFromDisk(self):
        """
        Get previously saved installation options from disk
        """
        self.LoadFromJson(GetJsonSettings(Name=self.__save_file_name, Scope="InstallationOptions"))
        
    def __str__(self) -> str:
        str = f"<InstallationOptions: SkipHashCheck={self.SkipHashCheck};"
        str += f"InteractiveInstallation={self.InteractiveInstallation};"
        str += f"RunAsAdministrator={self.RunAsAdministrator};"
        str += f"Version={self.Version};"
        str += f"Architecture={self.Architecture};"
        str += f"InstallationScope={self.InstallationScope};"
        str += f"CustomParameters={self.CustomParameters};"
        str += f"RemoveDataOnUninstall={self.RemoveDataOnUninstall}>"
        return str

class ManagerSource:
    
    class Capabilities:
        KnowsUpdateDate: bool = False
        KnowsPackageCount: bool = False
    
    Manager: 'PackageManagerWithSources' = None
    Name: str = ""
    Url: str = ""
    PackageCount: int = 0
    UpdateDate: str = ""
    
    def __init__(self, Manager: 'PackageManagerWithSources', Name: str, Url: str, PackageCount: int = 0, UpdateDate: str = ""):
        self.Manager = Manager
        self.Name = Name
        self.Url = Url
        self.PackageCount = PackageCount
        self.UpdateDate = UpdateDate
    

class PackageManagerCapabilities():
    CanRunAsAdmin: bool = False
    CanSkipIntegrityChecks: bool = False
    CanRunInteractively: bool = False
    CanRemoveDataOnUninstall: bool = False
    SupportsCustomVersions: bool = False
    SupportsCustomArchitectures: bool = False
    SupportsCustomScopes: bool = False
    SupportsPreRelease: bool = False
    SupportsCustomLocations: bool = False
    SupportsCustomSources: bool = False
    Sources: ManagerSource.Capabilities
    
    def __init__(self):
        self.Sources = ManagerSource.Capabilities()


class InstallationWidgetType(QWidget):
    finishInstallation: Signal
    addInfoLine: Signal
    counterSignal: Signal

    def __init__(self) -> None:
        raise RuntimeError("This class is a type declaration!")


class PackageManagerModule():
    NAME: str
    EXECUTABLE: str
    IconPath: str
    
    Capabilities: PackageManagerCapabilities
    LoadedIcons: bool
    Icon: QIcon
    icon: QIcon
    
    BLACKLISTED_PACKAGE_NAMES: list[str]
    BLACKLISTED_PACKAGE_IDS: list[str]
    BLACKLISTED_PACKAGE_VERSIONS: list[str]
    
    def __init__(self):
        self.Capabilities = PackageManagerCapabilities()
        self.BLACKLISTED_PACKAGE_NAMES: list[str] = []
        self.BLACKLISTED_PACKAGE_IDS: list[str] = []
        self.BLACKLISTED_PACKAGE_VERSIONS: list[str] = []
        self.LoadedIcons = None
        self.Icon = None
        self.icon = None
        self.IconPath = ""

    def isEnabled() -> bool:
        pass

    def getAvailableUpdates(self) -> list[UpgradablePackage]:
        """
        Will retieve the upgradable packages by the package manager in the format of a list[UpgradablePackage] object.
        """

    def getInstalledPackages(self) -> list[Package]:
        """
        Will retieve the intalled packages by the package manager in the format of a list[Package] object.
        """

    def getIcon(self, source: str) -> QIcon:
        """
        Will return the corresponding icon to the given source
        """

    def getPackageDetails(self, package: Package):
        """
        Will return a PackageDetails object containing the information of the given Package object
        """

    def startInstallation(self, package: Package, options: InstallationOptions, installationWidget: InstallationWidgetType) -> subprocess.Popen:
        """
        Starts a thread that installs the specified Package, making use of the given options. Reports the progress through the given InstallationWidget
        """

    def startUpdate(self, package: Package, options: InstallationOptions, installationWidget: InstallationWidgetType) -> subprocess.Popen:
        """
        Starts a thread that updates the specified Package, making use of the given options. Reports the progress through the given InstallationWidget
        """

    def installationThread(self, p: subprocess.Popen, options: InstallationOptions, installationWidget: InstallationWidgetType):
        """
        Internal method that handles the installation of the given package
        """

    def startUninstallation(self, package: Package, options: InstallationOptions, installationWidget: InstallationWidgetType) -> subprocess.Popen:
        """
        Starts a thread that removes the specified Package, making use of the given options. Reports the progress through the given InstallationWidget
        """

    def uninstallationThread(self, p: subprocess.Popen, options: InstallationOptions, installationWidget: InstallationWidgetType):
        """
        Internal method that handles the removal of the given package
        """

    def getParameters(self, options: InstallationOptions) -> list[str]:
        """
        Returns the list of parameters that the package manager ib nasis of the given InstallationOptions object
        """

    def detectManager(self, signal: Signal = None) -> None:
        """
        Detect if the package manager components exist.
        """

    def updateSources(self, signal: Signal = None) -> None:
        """
        Force update package manager's sources
        """

    def getPackagesForQuery(self, query: str) -> list[Package]:
        f"""
        Will retieve the packages for the given "query: str" from the package manager {self.NAME} in the format of a list[Package] object.
        """

class PackageManagerWithSources(PackageManagerModule):
    
    Sources: list[ManagerSource]
    KnownSources: list[ManagerSource]
    
    def __init__(self):
        super().__init__()
        self.Sources = []
        self.KnownSources = []
    
    def getSources(self) -> list[ManagerSource]:
        pass
    
    def installSource(self, source: ManagerSource, options: InstallationOptions, installationWidget: InstallationWidgetType) -> subprocess.Popen:
        pass
    
    def uninstallSource(self, source: ManagerSource, options: InstallationOptions, installationWidget: InstallationWidgetType) -> subprocess.Popen:
        pass
    


RETURNCODE_OPERATION_SUCCEEDED = 0
RETURNCODE_NO_APPLICABLE_UPDATE_FOUND = 92849
RETURNCODE_NEEDS_RESTART = 3

LIST_RETURNCODES_OPERATION_SUCCEEDED = (RETURNCODE_OPERATION_SUCCEEDED, RETURNCODE_NO_APPLICABLE_UPDATE_FOUND, RETURNCODE_NEEDS_RESTART)

RETURNCODE_FAILED = 1
RETURNCODE_INCORRECT_HASH = 2
RETURNCODE_NEEDS_ELEVATION = 1603
RETURNCODE_NEEDS_SCOOP_ELEVATION = -200
RETURNCODE_NEEDS_PIP_ELEVATION = -100

File: src/wingetui-interface/PackageEngine/PackageClasses.cs
using ModernWindow.Data;
using ModernWindow.Structures;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading.Tasks;

namespace ModernWindow.PackageEngine
{
    public enum PackageScope
    {
        // Repeated entries for coherence with Package Managers
        Global = 1,
        Machine = 1,
        Local = 0,
        User = 0,
    }
    public class Package : INotifyPropertyChanged
    {
        public class __serializable_updates_options_v1
        {
            public bool UpdatesIgnored { get; set; } = false;
            public string IgnoredVersion { get; set; } = "";
            public static async Task<__serializable_updates_options_v1> FromPackage(Package package)
            {
                var Serializable = new __serializable_updates_options_v1();
                Serializable.UpdatesIgnored = await package.HasUpdatesIgnored();
                Serializable.IgnoredVersion = await package.GetIgnoredUpdatesVersion();
                return Serializable;
            }
        }
        public class __serializable_bundled_package_v1
        {
            public string Id { get; set; } = "";
            public string Name { get; set; } = "";
            public string Version { get; set; } = "";
            public string Source { get; set; } = "";
            public string ManagerName { get; set; } = "";
            public InstallationOptions.__serializable_options_v1 InstallationOptions { get; set; }
            public __serializable_updates_options_v1 Updates { get; set; }
        }

        public class __serializable_incompatible_package_v1
        {
            public string Id { get; set; } = "";
            public string Name { get; set; } = "";
            public string Version { get; set; } = "";
            public string Source { get; set; } = "";
        }


        public AppTools bindings = AppTools.Instance;

        private bool __is_checked = false;

        public bool IsChecked { get { return __is_checked; } set { __is_checked = value; OnPropertyChanged(); } }
        public string IsCheckedAsString { get { return IsChecked ? "True" : "False"; } }
        public string Name { get; }
        public string Id { get; set; }
        public string Version { get; }
        public float VersionAsFloat { get; }
        public ManagerSource Source { get; set; }
        public PackageManager Manager { get; }
        public string UniqueId { get; }
        public string NewVersion { get; }
        public bool IsUpgradable { get; } = false;

        public event PropertyChangedEventHandler PropertyChanged;

        public PackageScope Scope { get; set; }
        public string SourceAsString
        {
            get
            {
                if (Source != null)
                    return Source.ToString();
                else return "";
            }
        }



        public Package(string name, string id, string version, ManagerSource source, PackageManager manager, PackageScope scope = PackageScope.Local)
        {
            Name = name;
            Id = id;
            Version = version;
            Source = source;
            Manager = manager;
            Scope = scope;
            UniqueId = $"{Manager.Properties.Name}\\{Id}\\{Version}";
            NewVersion = "";
            VersionAsFloat = GetFloatVersion();
        }

        public override bool Equals(object obj)
        {
            if (!(obj is Package))
                return false;
            else
                return Source == (obj as Package).Source && Id == (obj as Package).Id;
        }

        public Package _get_self_package()
        {
            return this;
        }

        public string GetIconId()
        {
            string iconId = Id.ToLower();
            if (Manager == bindings.App.Winget)
                iconId = String.Join('.', iconId.Split(".")[1..]);
            else if (Manager == bindings.App.Choco)
                iconId = iconId.Replace(".install", "").Replace(".portable", "");
            else if (Manager == bindings.App.Scoop)
                iconId = iconId.Replace(".app", "");
            return iconId;
        }

        public Uri GetIconUrl()
        {
            string iconId = GetIconId();
            if (CoreData.IconDatabaseData.ContainsKey(iconId))
                if (CoreData.IconDatabaseData[iconId].icon != "")
                    return new Uri(CoreData.IconDatabaseData[iconId].icon);

            return new Uri("ms-appx:///Assets/Images/package_color.png"); // TODO: Fallback image!
        }

        public float GetFloatVersion()
        {
            string _ver = "";
            bool _dotAdded = false;
            foreach (char _char in Version)
            {
                if (Char.IsDigit(_char))
                    _ver += _char;
                else if (_char == '.')
                {
                    if (!_dotAdded)
                    {
                        _ver += _char;
                        _dotAdded = true;
                    }
                }
            }
            float res = 0.0F;
            if (_ver != "" && _ver != ".")
                try
                {
                    return float.Parse(_ver);
                }
                catch { }
            return res;
        }

        public async Task AddToIgnoredUpdates(string version = "*")
        {
            string IgnoredId = $"{Manager.Properties.Name.ToLower()}\\{Id}";
            JsonObject IgnoredUpdatesJson = JsonNode.Parse(await File.ReadAllTextAsync(CoreData.IgnoredUpdatesDatabaseFile)) as JsonObject;
            if (IgnoredUpdatesJson.ContainsKey(IgnoredId))
                IgnoredUpdatesJson.Remove(IgnoredId);
            IgnoredUpdatesJson.Add(IgnoredId, version);
            await File.WriteAllTextAsync(CoreData.IgnoredUpdatesDatabaseFile, IgnoredUpdatesJson.ToString());
            bindings.App.mainWindow.NavigationPage.UpdatesPage.RemoveCorrespondingPackages(this);

            // TODO: Change InstalledPackages flag to show that the package is ignored, add to IgnoredPackages if applicable
        }

        public async Task RemoveFromIgnoredUpdates()
        {
            string IgnoredId = $"{Manager.Properties.Name.ToLower()}\\{Id}";
            JsonObject IgnoredUpdatesJson = JsonNode.Parse(await File.ReadAllTextAsync(CoreData.IgnoredUpdatesDatabaseFile)) as JsonObject;
            if (IgnoredUpdatesJson.ContainsKey(IgnoredId))
            {
                IgnoredUpdatesJson.Remove(IgnoredId);
                await File.WriteAllTextAsync(CoreData.IgnoredUpdatesDatabaseFile, IgnoredUpdatesJson.ToString());
            }

            // TODO: Change InstalledPackages flag to show that the package is no longer ignored

        }

        public async Task<bool> HasUpdatesIgnored(string version = "*")
        {
            string IgnoredId = $"{Manager.Properties.Name.ToLower()}\\{Id}";
            JsonObject IgnoredUpdatesJson = JsonNode.Parse(await File.ReadAllTextAsync(CoreData.IgnoredUpdatesDatabaseFile)) as JsonObject;
            if (IgnoredUpdatesJson.ContainsKey(IgnoredId) && (IgnoredUpdatesJson[IgnoredId].ToString() == "*" || IgnoredUpdatesJson[IgnoredId].ToString() == version))
                return true;
            else
                return false;
        }

        public async Task<string> GetIgnoredUpdatesVersion()
        {
            string IgnoredId = $"{Manager.Properties.Name.ToLower()}\\{Id}";
            JsonObject IgnoredUpdatesJson = JsonNode.Parse(await File.ReadAllTextAsync(CoreData.IgnoredUpdatesDatabaseFile)) as JsonObject;
            if (IgnoredUpdatesJson.ContainsKey(IgnoredId))
                return IgnoredUpdatesJson[IgnoredId].ToString();
            else
                return "";
        }

        protected void OnPropertyChanged([CallerMemberName] string name = null)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
        }

        public async Task<__serializable_bundled_package_v1> AsSerializable_BundledPackage()
        {
            var Serializable = new __serializable_bundled_package_v1();
            Serializable.Id = Id;
            Serializable.Name = Name;
            Serializable.Version = Version;
            Serializable.Source = Source.Name;
            Serializable.ManagerName = Manager.Name;
            var options = new InstallationOptions(this);
            Serializable.InstallationOptions = options.AsSerializable();
            var updates = await __serializable_updates_options_v1.FromPackage(this);
            Serializable.Updates = updates;
            return Serializable;
        }

        public __serializable_incompatible_package_v1 AsSerializable_IncompatiblePackage()
        {
            var Serializable = new __serializable_incompatible_package_v1();
            Serializable.Id = Id;
            Serializable.Name = Name;
            Serializable.Version = Version;
            Serializable.Source = Source.Name;
            return Serializable;
        }

    }

    public class UpgradablePackage : Package
    {
        public new string NewVersion { get; }

        public float NewVersionAsFloat { get; }
        public new bool IsUpgradable { get; } = true;

        public UpgradablePackage(string name, string id, string installed_version, string new_version, ManagerSource source, PackageManager manager, PackageScope scope = PackageScope.Local) : base(name, id, installed_version, source, manager, scope)
        {
            NewVersion = new_version;
            IsChecked = true;
            NewVersionAsFloat = GetFloatNewVersion();
        }

        public float GetFloatNewVersion()
        {
            string _ver = "";
            bool _dotAdded = false;
            foreach (char _char in NewVersion)
            {
                if (Char.IsDigit(_char))
                    _ver += _char;
                else if (_char == '.')
                {
                    if (!_dotAdded)
                    {
                        _ver += _char;
                        _dotAdded = true;
                    }
                }
            }
            float res = 0.0F;
            if (_ver != "" && _ver != ".")
                try
                {
                    return float.Parse(_ver);
                }
                catch (Exception)
                {
                }
            return res;
        }

        public new UpgradablePackage _get_self_package()
        {
            return this;
        }

        public bool NewVersionIsInstalled()
        {
            foreach (Package package in bindings.App.mainWindow.NavigationPage.InstalledPage.Packages)
                if (package.Manager == Manager && package.Id == Id && package.Version == NewVersion && package.Source.Name == Source.Name)
                    return true;
            return false;
        }
    }

    public class PackageDetails
    {
        public Package Package { get; }
        public string Name { get; }
        public string Id { get; }
        public string Version { get; }
        public string NewVersion { get; }
        public ManagerSource Source { get; }
        public string Description { get; set; } = "";
        public string Publisher { get; set; } = "";
        public string Author { get; set; } = "";
        public Uri? HomepageUrl { get; set; } = null;
        public string License { get; set; } = "";
        public Uri? LicenseUrl { get; set; } = null;
        public Uri? InstallerUrl { get; set; } = null;
        public string InstallerHash { get; set; } = "";
        public string InstallerType { get; set; } = "";
        public double InstallerSize { get; set; } = 0; // In Megabytes
        public Uri? ManifestUrl { get; set; } = null;
        public string UpdateDate { get; set; } = "";
        public string ReleaseNotes { get; set; } = "";
        public Uri? ReleaseNotesUrl { get; set; } = null;
        public string[] Versions { get; set; } = new string[0];
        public string[] Architectures { get; set; } = new string[0];
        public string[] Scopes { get; set; } = new string[0];
        public string[] Tags { get; set; } = new string[0];

        public PackageDetails(Package package)
        {
            Package = package;
            Name = package.Name;
            Id = package.Id;
            Version = package.Version;
            Source = package.Source;
            if (package is UpgradablePackage)
                NewVersion = ((UpgradablePackage)package).NewVersion;
            else
                NewVersion = "";
        }
    }

    public class InstallationOptions
    {
        public class __serializable_options_v1
        {
            public bool SkipHashCheck { get; set; } = false;
            public bool InteractiveInstallation { get; set; } = false;
            public bool RunAsAdministrator { get; set; } = false;
            public string Architecture { get; set; } = "";
            public string InstallationScope { get; set; } = "";
            public List<string> CustomParameters { get; set; }
            public bool PreRelease { get; set; } = false;
            public string CustomInstallLocation { get; set; } = "";
            public string Version { get; set; } = "";
        }

        public bool SkipHashCheck { get; set; } = false;
        public bool InteractiveInstallation { get; set; } = false;
        public bool RunAsAdministrator { get; set; } = false;
        public string Version { get; set; } = "";
        public Architecture? Architecture { get; set; } = null;
        public PackageScope? InstallationScope { get; set; } = null;
        public List<string> CustomParameters { get; set; } = new List<string>();
        public bool RemoveDataOnUninstall { get; set; } = false;
        public bool PreRelease { get; set; } = false;
        public string CustomInstallLocation { get; set; } = "";

        public Package Package { get; }

        private string _saveFileName = "Unknown.Unknown.InstallationOptions";

        public InstallationOptions(Package package, bool reset = false)
        {
            Package = package;
            _saveFileName = Package.Manager.Name.Replace(" ", "").Replace(".", "") + "." + Package.Id;
            if (!reset)
            {
                LoadOptionsFromDisk();
            }
        }

        public InstallationOptions(UpgradablePackage package, bool reset = false) : this((Package)package, reset)
        {
            if (!reset)
                LoadOptionsFromDisk();
        }

        public void LoadFromJsonString(string JSON)
        {
            __serializable_options_v1 options = JsonSerializer.Deserialize<__serializable_options_v1>(JSON);
            SkipHashCheck = options.SkipHashCheck;
            InteractiveInstallation = options.InteractiveInstallation;
            RunAsAdministrator = options.RunAsAdministrator;
            CustomInstallLocation = options.CustomInstallLocation;
            Version = options.Version;
            PreRelease = options.PreRelease;
            if (options.Architecture != "" && CommonTranslations.InvertedArchNames.ContainsKey(options.Architecture))
                Architecture = CommonTranslations.InvertedArchNames[options.Architecture];
            if (options.InstallationScope != "" && CommonTranslations.InvertedScopeNames_NonLang.ContainsKey(options.InstallationScope))
                InstallationScope = CommonTranslations.InvertedScopeNames_NonLang[options.InstallationScope];
            CustomParameters = options.CustomParameters;
        }

        public string GetJsonString()
        {
            return JsonSerializer.Serialize(AsSerializable());
        }

        public __serializable_options_v1 AsSerializable()
        {
            __serializable_options_v1 options = new();
            options.SkipHashCheck = SkipHashCheck;
            options.InteractiveInstallation = InteractiveInstallation;
            options.RunAsAdministrator = RunAsAdministrator;
            options.CustomInstallLocation = CustomInstallLocation;
            options.PreRelease = PreRelease;
            options.Version = Version;
            if (Architecture != null)
                options.Architecture = CommonTranslations.ArchNames[Architecture.Value];
            if (InstallationScope != null)
                options.InstallationScope = CommonTranslations.ScopeNames_NonLang[InstallationScope.Value];
            options.CustomParameters = CustomParameters;
            return options;
        }

        public void SaveOptionsToDisk()
        {
            try
            {
                string JSON = GetJsonString();
                string filename = Path.Join(CoreData.WingetUIInstallationOptionsDirectory, Package.Manager.Name + "." + Package.Id + ".json");
                File.WriteAllText(filename, JSON);
            }
            catch (Exception ex)
            {
                AppTools.Log(ex);
            }
        }

        public void LoadOptionsFromDisk()
        {
            try
            {
                string filename = Path.Join(CoreData.WingetUIInstallationOptionsDirectory, Package.Manager.Name + "." + Package.Id + ".json");
                if (!File.Exists(filename))
                    return;
                LoadFromJsonString(File.ReadAllText(filename));
            }
            catch (Exception e)
            {
                AppTools.Log(e.ToString());
            }
        }

        public override string ToString()
        {
            string customparams = CustomParameters != null ? string.Join(",", CustomParameters) : "[]";
            return $"<InstallationOptions: SkipHashCheck={SkipHashCheck};" +
                   $"InteractiveInstallation={InteractiveInstallation};" +
                   $"RunAsAdministrator={RunAsAdministrator};" +
                   $"Version={Version};" +
                   $"Architecture={Architecture};" +
                   $"InstallationScope={InstallationScope};" +
                   $"InstallationScope={CustomInstallLocation};" +
                   $"CustomParameters={customparams};" +
                   $"RemoveDataOnUninstall={RemoveDataOnUninstall}>";
        }
    }
}
